semgrep-core/parsing/Parse_rust_tree_sitter.ml

```diff
@@ -339,2122 +389,2262 @@ let map_macro_rule (env : env) ((v1, v2, v3) : CST.macro_rule) =
+and map_expression_ending_with_block (env : env) (x : CST.expression_ending_with_block): G.expr =
+  let map_loop_label_ (v1, v2) =
+    let loop_label = map_loop_label env v1 in
+    let colon = token env v2 (* ":" *) in
+    loop_label
+  in
   (match x with
-  | `Unsafe_blk (v1, v2) ->
-      let v1 = token env v1 (* "unsafe" *) in
-      let v2 = map_block env v2 in
-      todo env (v1, v2)
-  | `Async_blk (v1, v2, v3) ->
-      let v1 = token env v1 (* "async" *) in
-      let v2 =
-        (match v2 with
-        | Some tok -> token env tok (* "move" *)
-        | None -> todo env ())
-      in
-      let v3 = map_block env v3 in
-      todo env (v1, v2, v3)
-  | `Blk x -> map_block env x
-  | `If_exp x -> map_if_expression env x
-  | `If_let_exp x -> map_if_let_expression env x
-  | `Match_exp (v1, v2, v3) ->
-      let v1 = token env v1 (* "match" *) in
-      let v2 = map_expression env v2 in
-      let v3 = map_match_block env v3 in
-      todo env (v1, v2, v3)
-  | `While_exp (v1, v2, v3, v4) ->
-      let v1 =
-        (match v1 with
-        | Some (v1, v2) ->
-            let v1 = map_loop_label env v1 in
-            let v2 = token env v2 (* ":" *) in
-            todo env (v1, v2)
-        | None -> todo env ())
-      in
-      let v2 = token env v2 (* "while" *) in
-      let v3 = map_expression env v3 in
-      let v4 = map_block env v4 in
-      todo env (v1, v2, v3, v4)
-  | `While_let_exp (v1, v2, v3, v4, v5, v6, v7) ->
-      let v1 =
-        (match v1 with
-        | Some (v1, v2) ->
-            let v1 = map_loop_label env v1 in
-            let v2 = token env v2 (* ":" *) in
-            todo env (v1, v2)
-        | None -> todo env ())
-      in
-      let v2 = token env v2 (* "while" *) in
-      let v3 = token env v3 (* "let" *) in
-      let v4 = map_pattern env v4 in
-      let v5 = token env v5 (* "=" *) in
-      let v6 = map_expression env v6 in
-      let v7 = map_block env v7 in
-      todo env (v1, v2, v3, v4, v5, v6, v7)
-  | `Loop_exp (v1, v2, v3) ->
-      let v1 =
-        (match v1 with
-        | Some (v1, v2) ->
-            let v1 = map_loop_label env v1 in
-            let v2 = token env v2 (* ":" *) in
-            todo env (v1, v2)
-        | None -> todo env ())
-      in
-      let v2 = token env v2 (* "loop" *) in
-      let v3 = map_block env v3 in
-      todo env (v1, v2, v3)
-  | `For_exp (v1, v2, v3, v4, v5, v6) ->
-      let v1 =
-        (match v1 with
-        | Some (v1, v2) ->
-            let v1 = map_loop_label env v1 in
-            let v2 = token env v2 (* ":" *) in
-            todo env (v1, v2)
-        | None -> todo env ())
-      in
-      let v2 = token env v2 (* "for" *) in
-      let v3 = map_pattern env v3 in
-      let v4 = token env v4 (* "in" *) in
-      let v5 = map_expression env v5 in
-      let v6 = map_block env v6 in
-      todo env (v1, v2, v3, v4, v5, v6)
-  | `Const_blk x -> map_const_block env x
+   | `Unsafe_blk (v1, v2) ->
```

Comment:
You can use OtherStmtWithStmt and add a new UnsafeBlock constructor in
other_stmt_with_stmt_operator type.